### PR TITLE
Merge #211 to master

### PR DESCRIPTION
**Pull request by LazoYoung**
I'm requesting to merge the commits from this branch into your master.

Issues resolved in this branch:
- #211
- #213 

Known issues arisen in this branch:
- Curved shot is broken (which is solved in the next PR)

### CODE REVIEW

File: blockball-api/src/main/java/com/github/shynixn/blockball/api/persistence/entity/BallMeta.kt [hunk 1]
```diff
@@ -64,6 +64,12 @@ interface BallMeta {
      * Should the ball always bounce of walls?
      */
     var alwaysBouce: Boolean
+
```
Comment (Shynixn) [inline blockball-api/src/main/java/com/github/shynixn/blockball/api/persistence/entity/BallMeta.kt]: Is this relevant to the arena creator?

File: blockball-bukkit-plugin/blockball-bukkit-nms-114R1/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/nms/v1_14_R1/BallDesign.kt [hunk 2]
```diff
@@ -49,12 +52,12 @@ import java.util.*
  */
 class BallDesign(location: Location, ballMeta: BallMeta, persistent: Boolean, uuid: UUID = UUID.randomUUID(), owner: LivingEntity?) :
     EntityArmorStand((location.world as CraftWorld).handle, location.x, location.y, location.z), NMSBallProxy {
-    private val itemService = BlockBallApi.resolve(ItemService::class.java)
```
Comment (Shynixn) [inline blockball-bukkit-plugin/blockball-bukkit-nms-114R1/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/nms/v1_14_R1/BallDesign.kt]: These changes are too complex and need to be tested by me in detail.

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/BlockBallDependencyInjectionBinder.kt [hunk 3]
```diff
@@ -93,6 +97,19 @@ class BlockBallDependencyInjectionBinder(private val plugin: BlockBallPlugin) :
         bind(CommandService::class.java).to(CommandServiceImpl::class.java).`in`(Scopes.SINGLETON)
         bind(GameExecutionService::class.java).to(GameExecutionServiceImpl::class.java).`in`(Scopes.SINGLETON)
 
+        when {
+            // TODO Implement EntityRegistrationServiceImpl lower than 1.14
```
Comment (Shynixn) [inline blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/BlockBallDependencyInjectionBinder.kt]: Why is the TODO still present?

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/listener/BallListener.kt [hunk 4]
```diff
@@ -65,7 +64,7 @@ class BallListener @Inject constructor(
      */
     @EventHandler
     fun onChunkSaveEvent(event: ChunkUnloadEvent) {
-        event.chunk.entities.filter { e -> e is ArmorStand }
+        event.chunk.entities.filter { e -> e.customName == "ResourceBallsPlugin" }
```
Comment (Shynixn) [inline blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/listener/BallListener.kt]: CustomName is a dangerous method which can return null. This has to get solved differently.

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt [hunk 5]
```diff
@@ -67,51 +76,72 @@ class BallEntityServiceImpl @Inject constructor(private val pluginProxy: PluginP
...
+    override fun registerEntitiesOnServer(): Boolean {
+        if (registered) {
+            return false
+        }
+
+        val slimeClazz = findClazz("com.github.shynixn.blockball.bukkit.logic.business.nms.VERSION.BallHitBox", pluginProxy)
+        val armorStandClazz = findClazz("com.github.shynixn.blockball.bukkit.logic.business.nms.VERSION.BallDesign", pluginProxy)
+
+        entityRegistry.register(slimeClazz, EntityType.SLIME)
+        entityRegistry.register(armorStandClazz, EntityType.ARMORSTAND)
```
Comment (Shynixn) [inline blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt]: Do not register the Armorstand! This is very dangerous and can cause chunk corruption. 

I know that PetBlocks contains the EntityType.Armorstand field in the enum, however it is unused.

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt [hunk 6]
```diff
@@ -67,51 +76,72 @@ class BallEntityServiceImpl @Inject constructor(private val pluginProxy: PluginP
...
-
-                        try{
-                            (entity as Any).javaClass.getDeclaredMethod("deleteFromWorld").invoke(entity)
-                        }catch(e : Exception){
-                        }
+            if (entity !is LivingEntity) {
+                continue
+            }
+
+            if (entity.customName == "ResourceBallsPlugin") {
```
Comment (Shynixn) [inline blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt]: Check if entity.customName is not null.

File: blockball-bukkit-plugin/src/test/java/unittest/BallListenerTest.kt [hunk 7]
```diff
@@ -11,8 +11,8 @@ import com.github.shynixn.blockball.api.persistence.entity.Particle
 import com.github.shynixn.blockball.api.persistence.entity.Sound
 import com.github.shynixn.blockball.bukkit.logic.business.listener.BallListener
 import org.bukkit.Chunk
-import org.bukkit.entity.ArmorStand
 import org.bukkit.entity.Entity
+import org.bukkit.entity.Slime
 import org.bukkit.event.world.ChunkLoadEvent
 import org.bukkit.event.world.ChunkUnloadEvent
 import org.junit.jupiter.api.Assertions
```
Comment (Shynixn) [inline blockball-bukkit-plugin/src/test/java/unittest/BallListenerTest.kt]: Good job on updating the tests. I did not expect that to see.

File: build.gradle [hunk 8]
```diff
@@ -260,7 +261,8 @@ task downloadDependencies(type: Exec) {
     text += " && mvn install:install-file -Dfile=spigot-1.12.jar -DgroupId=org.spigotmc -DartifactId=spigot112R1 -Dversion=1.12.0-R1.0 -Dpackaging=jar"
     text += " && mvn install:install-file -Dfile=spigot-1.13.jar -DgroupId=org.spigotmc -DartifactId=spigot113R1 -Dversion=1.13.0-R1.0 -Dpackaging=jar"
     text += " && mvn install:install-file -Dfile=spigot-1.13.2.jar -DgroupId=org.spigotmc -DartifactId=spigot113R2 -Dversion=1.13.2-R2.0 -Dpackaging=jar"
-    text += " && mvn install:install-file -Dfile=spigot-1.14.2.jar -DgroupId=org.spigotmc -DartifactId=spigot114R1 -Dversion=1.14.1-R1.0 -Dpackaging=jar"
+    text += " && mvn install:install-file -Dfile=spigot-1.14.2.jar -DgroupId=org.spigotmc -DartifactId=spigot114R1 -Dversion=1.14.2-R1.0 -Dpackaging=jar"
```
Comment (Shynixn) [inline build.gradle]: Remove the the installing of 1.14.2. We do only need 1.14.4.

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/listener/BallListener.kt (hunk 4 above)
Comment (Shynixn) [reply blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/listener/BallListener.kt]: Does the "e.customName == "ResourceBallsPlugin" return false or throw a NullPointException in Kotlin? I am actually not sure right now.

File: blockball-api/src/main/java/com/github/shynixn/blockball/api/persistence/entity/BallMeta.kt (hunk 1 above)
Comment (LazoYoung) [reply blockball-api/src/main/java/com/github/shynixn/blockball/api/persistence/entity/BallMeta.kt]: Probably not. Would you tell me the reason we have to skip interaction checks? And how much amount of tick that you think is relevant?

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/BlockBallDependencyInjectionBinder.kt (hunk 3 above)
Comment (LazoYoung) [reply blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/BlockBallDependencyInjectionBinder.kt]: Looks like I omitted the commit for this PR. I'll fix it

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/listener/BallListener.kt (hunk 4 above)
Comment (LazoYoung) [reply blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/listener/BallListener.kt]: AFAIK equals check in Kotlin is null-safe.
https://kotlinlang.org/docs/reference/operator-overloading.html#equals

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt (hunk 5 above)
Comment (LazoYoung) [reply blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt]: I couldn't reckon that armorstand should never get registered. Okay

File: build.gradle (hunk 8 above)
Comment (LazoYoung) [reply build.gradle]: Okay I'll fix

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt (hunk 6 above)
Comment (LazoYoung) [reply blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/service/BallEntityServiceImpl.kt]: Like we discussed above, we don't need to check nullability for this

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/BlockBallDependencyInjectionBinder.kt (hunk 3 above)
Comment (LazoYoung) [reply blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/BlockBallDependencyInjectionBinder.kt]: [Checkout my latest commit.](https://github.com/LazoYoung/BlockBall/commit/b86683be4ab143a08375778b2687cda72e19e86f)

File: blockball-bukkit-plugin/src/test/java/unittest/BallListenerTest.kt (hunk 7 above)
Comment (LazoYoung) [reply blockball-bukkit-plugin/src/test/java/unittest/BallListenerTest.kt]: I don't know how unittest works, I'd like to leave this up to you.

File: blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/proxy/BallProxyImpl.kt [hunk 9]
```diff
@@ -244,7 +244,11 @@ class BallProxyImpl(
             return
         }
 
-        val vector = hitbox.location.toVector().subtract(entity.location.toVector()).normalize()
+        /*
+        val vector = design.location.toVector().subtract(entity.location.toVector()).normalize()
+            .multiply(meta.movementModifier.horizontalKickModifier)
+         */
+        val vector = entity.location.direction.normalize()
```
Comment (LazoYoung) [inline blockball-bukkit-plugin/src/main/java/com/github/shynixn/blockball/bukkit/logic/business/proxy/BallProxyImpl.kt]: Kick direction is now equal to the direction of the player
